File: renderers/web/tests/nineslice.test.js
// A picture cut into nine, which the browser draws as a border image.
//
// What matters here is that the frame is painted over the box rather than inside it. A border with a
// width of its own would push every child the engine placed inward by the thickness of the frame, and
// the same tree would come out laid out one way in a browser and another way on a phone.

import { install } from "../dom.js";
import { WebRenderer } from "../renderer.js";

const surface = install();
const renderer = new WebRenderer(surface, () => {});

function assert(condition, message) {
    if (!condition) {
        console.error(`failed: ${message}`);
        process.exit(1);
    }
}

function frame(id, props) {
    renderer.apply([
        { op: "create", id, type: "nineslice", props },
        { op: "insert", id, parent: 0, index: 1 },
    ]);

    return renderer.nodes.get(id).element;
}

// The cuts are written into the slice in the picture's own pixels, and into the width in points.
{
    const element = frame(1, {
        source: "/assets/window.png",
        slice: { top: 10, right: 20, bottom: 30, left: 40 },
        sliceScale: 2,
    });

    assert(element.style.borderImageSource === 'url("/assets/window.png")',
        `the picture is the source, got ${element.style.borderImageSource}`);

    assert(element.style.borderImageSlice === "10 20 30 40 fill",
        `the cuts are the picture's own pixels, and the middle is filled, got ${element.style.borderImageSlice}`);

    assert(element.style.borderImageWidth === "20px 40px 60px 80px",
        `how thick it comes out is the cuts drawn at the scale, got ${element.style.borderImageWidth}`);

    assert(element.style.borderImageRepeat === "stretch",
        `the edges stretch rather than repeating, got ${element.style.borderImageRepeat}`);
}

// The border itself takes no room, so nothing the engine placed inside one is moved by the frame.
{
    const element = frame(2, { source: "/assets/window.png", slice: { top: 8, right: 8, bottom: 8, left: 8 } });

    assert(element.style.borderWidth === "" || element.style.borderWidth === "0px",
        `a frame takes no room of its own, got ${element.style.borderWidth}`);
}

// Drawn at nothing in particular, one point is one pixel of the picture.
{
    const element = frame(3, { source: "/assets/button.png", slice: { top: 14, right: 14, bottom: 14, left: 14 } });

    assert(element.style.borderImageWidth === "14px 14px 14px 14px",
        `with no scale the border is the cuts themselves, got ${element.style.borderImageWidth}`);
}

// A cut that arrives before the picture, or a picture before the cuts, draws nothing rather than half.
{
    const element = frame(4, { slice: { top: 6, right: 6, bottom: 6, left: 6 } });

    assert(element.style.borderImageSource === "",
        `a frame with no picture paints nothing, got ${element.style.borderImageSource}`);

    renderer.apply([{ op: "update", id: 4, props: { source: "/assets/late.png" } }]);

    assert(element.style.borderImageSource === 'url("/assets/late.png")',
        "and paints as soon as the picture it was waiting for arrives");

    assert(element.style.borderImageSlice === "6 6 6 6 fill",
        "with the cuts it already had");
}

// The renderer says it draws one, so a screen can ask before it builds itself out of frames.
{
    assert(renderer.capabilities.nineSlice === true, "the browser declares that it draws a frame");
}

console.log("web.nineslice ok");

// A frame draws the artwork the state it is in names, and the plain one for a state that names none.
{
    const element = frame(9, {
        source: "/assets/plate.png",
        sources: { pressed: "/assets/plate-down.png", disabled: "/assets/plate-off.png" },
        slice: 8,
    });

    const node = renderer.nodes.get(9);

    assert(element.style.borderImageSource === 'url("/assets/plate.png")',
        `a frame nobody is touching draws the plain picture, got ${element.style.borderImageSource}`);

    element.dispatch("pointerdown", {});
    assert(element.style.borderImageSource === 'url("/assets/plate-down.png")',
        `a frame under a finger draws the pressed picture, got ${element.style.borderImageSource}`);

    element.dispatch("pointerup", {});
    assert(element.style.borderImageSource === 'url("/assets/plate.png")',
        `and comes back when the finger lifts, got ${element.style.borderImageSource}`);

    // A state the frame carries no artwork for is drawn with the plain picture rather than with nothing.
    element.dispatch("pointerenter", { pointerType: "mouse" });
    assert(element.style.borderImageSource === 'url("/assets/plate.png")',
        `a state with no artwork of its own keeps the plain picture, got ${element.style.borderImageSource}`);

    renderer.apply([{ op: "update", id: 9, props: { disabled: true } }]);
    assert(node.off === true, "the frame was told it may not be pressed");
    assert(element.style.borderImageSource === 'url("/assets/plate-off.png")',
        `a frame that may not be pressed draws the picture for that, got ${element.style.borderImageSource}`);
}

File: gui/tests/nineslice_test.lua
local async = require("async")
local fs = require("fs")
local gui = require("gui")
local bundle = require("gui.assets.bundle")
local natural = require("gui.layout.natural")

local ROOT = assert(os.getenv("VARN_TEST_DIR"), "VARN_TEST_DIR is not set")

local function start(description, options)
    local renderer = gui.headless()
    options = options or {}
    options.size = options.size or { width = 320, height = 480 }

    local runtime = gui.start(description, renderer, options)

    for _ = 1, 4 do
        if not runtime:needsCommit() then
            break
        end

        runtime:commit()
    end

    return runtime, renderer
end

local function refused(build, says)
    local ok, problem = pcall(build)

    assert(not ok, "must be refused: " .. says)
    assert(tostring(problem):find(says, 1, true) ~= nil,
        "the refusal must say " .. says .. ", said " .. tostring(problem))
end

async.run(function()
    -- One number cuts all four edges, and the renderer is never handed the shorthand.
    --
    -- A shorthand each renderer reads for itself is three chances to read it differently, so it is written
    -- the one way here and crosses as four numbers whichever way it was asked for.
    do
        local _, renderer = start(gui.NineSlice { source = "panel.png", slice = 12 })
        local slice = renderer:find("nineslice").props.slice

        assert(type(slice) == "table", "a slice crosses as four numbers, got a " .. type(slice))
        assert(slice.top == 12 and slice.right == 12 and slice.bottom == 12 and slice.left == 12,
            "one number cuts every edge")
    end

    -- Each edge may be cut on its own, which is what a window with a bar across its top needs.
    do
        local _, renderer = start(gui.NineSlice {
            source = "panel.png",
            slice = { top = 40, right = 20, bottom = 16, left = 20 },
        })

        local slice = renderer:find("nineslice").props.slice

        assert(slice.top == 40 and slice.right == 20 and slice.bottom == 16 and slice.left == 20,
            "the four edges cross as they were written")
    end

    -- What cannot be drawn is refused where it was written rather than drawn as nothing.
    do
        refused(function() return gui.NineSlice { slice = 8 } end, "needs a source")
        refused(function() return gui.NineSlice { source = "panel.png" } end, "needs a slice")

        refused(function() return gui.NineSlice { source = "panel.png", slice = { top = 8, right = 8 } } end,
            "needs a slice")

        refused(function() return gui.NineSlice { source = "panel.png", slice = -4 } end,
            "whole number of pixels")

        refused(function() return gui.NineSlice { source = "panel.png", slice = 2.5 } end,
            "whole number of pixels")

        refused(function() return gui.NineSlice { source = "panel.png", slice = 8, sliceScale = 0 } end,
            "above nothing")
    end

    -- A frame is worth at least the two corners it is cut at, since narrower than that it cannot be drawn.
    --
    -- The corners would overlap and what comes out is not the picture, so the minimum is not a number fixed
    -- for the type but one worked out from the cuts the node was given.
    do
        local least = natural.sizeOf("nineslice", { slice = { top = 10, right = 6, bottom = 14, left = 6 }, sliceScale = 1 })

        assert(least.minWidth == 12, "a frame is at least its two side cuts wide, got " .. tostring(least.minWidth))
        assert(least.minHeight == 24, "and its two end cuts tall, got " .. tostring(least.minHeight))

        local drawn = natural.sizeOf("nineslice", { slice = { top = 10, right = 6, bottom = 14, left = 6 }, sliceScale = 3 })

        assert(drawn.minWidth == 36, "drawn larger, the minimum grows with it, got " .. tostring(drawn.minWidth))
    end

    -- A frame narrower than its corners is laid out at the width it needs rather than at the one it was told.
    do
        local _, renderer = start(gui.View {
            style = { width = 200, height = 200 },
            gui.NineSlice { source = "panel.png", slice = 30, sliceScale = 2, style = { width = 20, height = 20 } },
        })

        local frame = renderer:find("nineslice")

        assert(frame.frame.width == 120, "the frame is as wide as its corners, got " .. tostring(frame.frame.width))
        assert(frame.frame.height == 120, "and as tall, got " .. tostring(frame.frame.height))
    end

    -- The picture a frame is cut from is the one that was named, never a denser variant of it.
    --
    -- The cuts point at that file's own pixels, so handing over an `@2x` twice the size would cut it at the
    -- same four numbers and every corner would come out half the width it was drawn at.
    do
        local project = ROOT .. "/framed"

        fs.mkdir(project .. "/assets/images"):await()
        fs.writeFile(project .. "/manifest.lua", "return { identifier = 'd.v.framed', version = '1.0.0' }"):await()
        fs.writeFile(project .. "/assets/images/panel.png", "the picture itself"):await()
        fs.writeFile(project .. "/assets/images/panel@2x.png", "twice the size"):await()
        fs.writeFile(project .. "/assets/images/badge.png", "a picture"):await()
        fs.writeFile(project .. "/assets/images/badge@2x.png", "twice the size"):await()

        local opened = bundle.openDirectory(project)

        local _, renderer = start(gui.View {
            gui.NineSlice { source = "panel.png", slice = 8 },
            gui.Image { source = "badge.png", style = { width = 40, height = 40 } },
        }, { assets = opened, scale = 2 })

        local frame = renderer:find("nineslice").props.source
        local picture = renderer:find("image").props.source

        assert(frame:find("panel.png", 1, true) ~= nil and frame:find("@2x", 1, true) == nil,
            "a frame is cut from the picture it named, got " .. frame)
        assert(picture:find("badge@2x.png", 1, true) ~= nil,
            "while a picture still takes the variant the surface is worth, got " .. picture)
    end

    -- A frame carries artwork per state, and each of those is cut from its own picture rather than a variant.
    --
    -- A frame is how a button is drawn from artwork, and a button is a different picture while a finger
    -- is on it, so the picture for each state is named the way the plain one is and resolved the same
    -- way — a denser variant would be cut at the same four numbers and come out half the width.
    do
        local project = ROOT .. "/stated"

        fs.mkdir(project .. "/assets/images"):await()
        fs.writeFile(project .. "/manifest.lua", "return { identifier = 'd.v.stated', version = '1.0.0' }"):await()

        for _, name in ipairs({ "plate", "plate-down", "plate-off" }) do
            fs.writeFile(project .. "/assets/images/" .. name .. ".png", "the picture itself"):await()
            fs.writeFile(project .. "/assets/images/" .. name .. "@2x.png", "twice the size"):await()
        end

        local _, renderer = start(gui.NineSlice {
            source = "plate.png",
            sources = { pressed = "plate-down.png", disabled = "plate-off.png" },
            slice = 8,
        }, { assets = bundle.openDirectory(project), scale = 2 })

        local held = renderer:find("nineslice").props.sources

        assert(type(held) == "table", "the artwork per state crosses as a table, got a " .. type(held))
        assert(held.pressed:find("plate-down.png", 1, true) ~= nil and held.pressed:find("@2x", 1, true) == nil,
            "a state is cut from the picture it named, got " .. tostring(held.pressed))
        assert(held.disabled:find("plate-off.png", 1, true) ~= nil,
            "and every state named is resolved, got " .. tostring(held.disabled))
        assert(held.hovered == nil, "a state with no artwork of its own is not invented")
    end

    -- A frame that names artwork for a state nobody has is refused where it was written.
    do
        refused(function()
            return gui.NineSlice { source = "panel.png", slice = 8, sources = { sideways = "no.png" } }
        end, "sideways")

        refused(function()
            return gui.NineSlice { source = "panel.png", slice = 8, sources = { pressed = 12 } }
        end, "sources.pressed")

        refused(function()
            return gui.NineSlice { source = "panel.png", slice = 8, sources = "one.png" }
        end, "sources names a picture per state")
    end

    -- A frame reports both edges of a press, which is what lets a button look pressed while it is.
    --
    -- A press is reported when the finger lifts, so a frame that only had that went down as it was let
    -- go: the whole of the animation happened after the press was over. A button drawn from artwork is
    -- the reason this component exists, and one that cannot show it is being held is not a button.
    do
        local told = {}
        local runtime, renderer = start(gui.NineSlice {
            source = "panel.png",
            slice = 8,
            onPressIn = function() told[#told + 1] = "in" end,
            onPressOut = function() told[#told + 1] = "out" end,
            onPress = function() told[#told + 1] = "press" end,
        })

        local frame = renderer:find("nineslice")

        runtime:dispatch(frame.id, "onPressIn", nil)
        runtime:dispatch(frame.id, "onPressOut", nil)
        runtime:dispatch(frame.id, "onPress", nil)

        assert(table.concat(told, ",") == "in,out,press",
            "a frame hears the finger land, lift and count, heard " .. table.concat(told, ","))
    end

    -- The cuts the sample writes are the cuts the artwork was drawn for.
    --
    -- A picture and the four numbers it is cut at belong together and live apart: the numbers are in the
    -- screen that draws the frame and the picture is drawn by a script that has to know them to check
    -- its own work. Two places holding the same four numbers agree with nothing unless something says
    -- so, and what a disagreement looks like is ornament pulled out of shape on a device.
    do
        local demo = fs.readFile("sample/demos/frames.lua"):await()
        local script = fs.readFile("tools/draw-frames.py"):await()

        local written = {}

        for source, slice in demo:gmatch('source = "([%w%-%.]+)", slice = ([^\n]-),?\n') do
            written[source] = slice:gsub("%s+", ""):gsub("[,}]+$", "}"):gsub("^([^{])[,}]*$", "%1")
        end

        assert(next(written) ~= nil, "the demo must name the pictures it cuts")

        for source in pairs(written) do
            assert(script:find('"' .. source .. '"', 1, true) ~= nil,
                "the sample cuts " .. source .. " and tools/draw-frames.py never draws it")
        end

        -- The one cut unevenly is the one worth naming, since it is the one a reader would get wrong.
        assert(written["window-titled.png"] == "{top=46,right=28,bottom=28,left=28}",
            "the titled window is cut deeper along its top, the demo says " .. tostring(written["window-titled.png"]))
        assert(script:find("(46, 28, 28, 28)", 1, true) ~= nil,
            "and the script that draws it must check itself against those same four numbers")
    end

    print("gui.nineslice ok")
end)
